### ISSUE TITLE
ArtifactGC cannot resolve template when using dependent WorkflowTemplates

### ISSUE THREAD

**Issue by sakvaka** [argoproj/argo-workflows#10158]
### Pre-requisites

- [X] I have double-checked my configuration
- [X] I can confirm the issues exists when I tested with `:latest`
- [ ] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/master/docs/CONTRIBUTING.md))

### What happened/what you expected to happen?

First of all, thanks for the fix in #10015 . Now I have encountered a new issue with ArtifactGC.

I have created two `WorkflowTemplate` manifests in my environment, one of which refers to the templates of the other. The dependent `WorkflowTemplate` defines an artifact. The independent `WorkflowTemplate` defines an `artifactGC` policy in its `spec` field, on the workflow level.

When I use Argo UI to submit a workflow based on the latter `WorkflowTemplate` and after completion delete the workflow, nothing happens, artifacts are not deleted and finalizers are not deleted. Looking at the logs, it seems that ArtifactGC cannot resolve the template that defines the artifact (`Error: template name <NAME> belonging to node <NODE> not found??`).

The same happens when I use the Argo-CLI to submit a workflow based on the `WorkflowTemplate`:

```bash
argo submit -n argo --from workflowtemplate/test-wft2
```
I would expect that ArtifactGC can resolve all artifacts that the workflow has created, no matter where the generation of the artifact was originally specified.

As a context: we have data processing tools defined as templates in different `WorkflowTemplate` manifests (say, `tool1`, `tool2` etc.), and also templates representing how to chain these tools together, stored in separate `WorkflowTemplate` (say, `tool-workflow1`, `tool-workflow2`). We also have templates that can execute many of these workflows in parallel (say, `parallel-tool-workflow1`). So in principle, Argo Workflows allows a very complicated nesting of templates coming from the different manifests in the cluster.

### Version

v. 3.4.4 (Helm chart 0.20.12)

### Paste a small workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
---
apiVersion: argoproj.io/v1alpha1
kind: WorkflowTemplate
metadata:
  name: test-wft
spec:
  entrypoint: main
  templates:
    - name: main
      container:
        image: argoproj/argosay:v2
        command:
          - sh
          - -c
        args:
          - |
            echo "can throw this away" > /tmp/temporary-artifact.txt
      outputs:
        artifacts:
          - name: temporary-artifact
            path: /tmp/temporary-artifact.txt
            s3:
              key: temporary-artifact.txt
---
apiVersion: argoproj.io/v1alpha1
kind: WorkflowTemplate
metadata:
  name: test-wft2
spec:
  entrypoint: steps-direct
  artifactGC:
    strategy: OnWorkflowDeletion
    serviceAccountName: artifactgc-sa
  
  templates:
    - name: steps-direct
      steps:
        - - name: main
            templateRef:
              name: test-wft
              template: main
```

### Logs from the workflow controller

```text
2022-12-02 11:32:44	
time="2022-12-02T09:32:44.542Z" level=info msg="DeleteCollection workflowtaskresults 200"
	
2022-12-02 11:32:44	
time="2022-12-02T09:32:44.537Z" level=info msg="Workflow update successful" namespace=argo phase=Succeeded resourceVersion=845496 workflow=test-wft2-cddwq
	
2022-12-02 11:32:44	
time="2022-12-02T09:32:44.536Z" level=info msg="Update workflows 200"
	
2022-12-02 11:32:44	
time="2022-12-02T09:32:44.512Z" level=info msg="Workflow to be dehydrated" Workflow Size=3533
	
2022-12-02 11:32:44	
time="2022-12-02T09:32:44.511Z" level=debug msg="Log changes patch: {\"status\":{\"artifactGCStatus\":{\"strategiesProcessed\":{\"OnWorkflowDeletion\":true}}}}"
	
2022-12-02 11:32:44	
time="2022-12-02T09:32:44.510Z" level=info msg="Checking daemoned children of " namespace=argo workflow=test-wft2-cddwq
	
2022-12-02 11:32:44	
time="2022-12-02T09:32:44.510Z" level=error msg="failed to GC artifacts" error="can't process Artifact GC Strategy OnWorkflowDeletion: template name \"main\" belonging to node {ID:test-wft2-cddwq-2556220558 Name:test-wft2-cddwq(0)[0].main(0) DisplayName:main(0) Type:Pod TemplateName: TemplateRef:&TemplateRef{Name:test-wft,Template:main,ClusterScope:false,} TemplateScope:local/ Phase:Succeeded BoundaryID:test-wft2-cddwq-2446534878 Message: StartedAt:2022-12-02 09:31:18 +0000 UTC FinishedAt:2022-12-02 09:32:23 +0000 UTC EstimatedDuration:0 Progress:1/1 ResourcesDuration:9s*(1 cpu),9s*(100Mi memory) PodIP: Daemoned:<nil> Inputs:nil Outputs:&Outputs{Parameters:[]Parameter{},Artifacts:[]Artifact{Artifact{Name:temporary-artifact,Path:/tmp/temporary-artifact.txt,Mode:nil,From:,ArtifactLocation:ArtifactLocation{ArchiveLogs:nil,S3:&S3Artifact{S3Bucket:S3Bucket{Endpoint:,Bucket:,Region:,Insecure:nil,AccessKeySecret:nil,SecretKeySecret:nil,RoleARN:,UseSDKCreds:false,CreateBucketIfNotPresent:nil,EncryptionOptions:nil,},Key:temporary-artifact.txt,},Git:nil,HTTP:nil,Artifactory:nil,HDFS:nil,Raw:nil,OSS:nil,GCS:nil,Azure:nil,},GlobalName:,Archive:nil,Optional:false,SubPath:,RecurseMode:false,FromExpression:,ArtifactGC:nil,Deleted:false,},},Result:nil,ExitCode:*0,} Children:[] OutboundNodes:[] HostNodeName:<REDACTED> MemoizationStatus:nil SynchronizationStatus:nil} not found??" namespace=argo workflow=test-wft2-cddwq
	
2022-12-02 11:32:44	
time="2022-12-02T09:32:44.510Z" level=debug msg="processing Artifact GC Strategy OnWorkflowDeletion" namespace=argo workflow=test-wft2-cddwq
	
2022-12-02 11:32:44	
time="2022-12-02T09:32:44.510Z" level=debug msg="processing Artifact GC Strategy OnWorkflowDeletion" namespace=argo workflow=test-wft2-cddwq
	
2022-12-02 11:32:44	
time="2022-12-02T09:32:44.509Z" level=info msg="Processing workflow" namespace=argo workflow=test-wft2-cddwq
```

### Logs from in your workflow's wait container

```text
2022-12-02 11:32:23	
time="2022-12-02T09:32:23.446Z" level=info msg="Alloc=6564 TotalAlloc=16064 Sys=22482 NumGC=5 Goroutines=9"
	
2022-12-02 11:32:23	
time="2022-12-02T09:32:23.442Z" level=info msg="Deadline monitor stopped"
	
2022-12-02 11:32:23	
time="2022-12-02T09:32:23.442Z" level=info msg="stopping progress monitor (context done)" error="context canceled"
	
2022-12-02 11:32:23	
time="2022-12-02T09:32:23.440Z" level=info msg="Create workflowtaskresults 201"
	
2022-12-02 11:32:23	
time="2022-12-02T09:32:23.372Z" level=info msg="Successfully saved file: /tmp/argo/outputs/artifacts/temporary-artifact.tgz"
	
2022-12-02 11:32:23	
time="2022-12-02T09:32:23.372Z" level=info msg="not deleting local artifact" localArtPath=/tmp/argo/outputs/artifacts/temporary-artifact.tgz
	
2022-12-02 11:32:23	
time="2022-12-02T09:32:23.372Z" level=info msg="Save artifact" artifactName=temporary-artifact duration=311.209393ms error="<nil>" key=temporary-artifact.txt
	
2022-12-02 11:32:23	
time="2022-12-02T09:32:23.220Z" level=info msg="Saving file to s3" bucket=<REDACTED> endpoint=s3.amazonaws.com key=temporary-artifact.txt path=/tmp/argo/outputs/artifacts/temporary-artifact.tgz
	
2022-12-02 11:32:23	
time="2022-12-02T09:32:23.061Z" level=info msg="Creating minio client using AWS SDK credentials"
	
2022-12-02 11:32:23	
time="2022-12-02T09:32:23.061Z" level=info msg="S3 Save path: /tmp/argo/outputs/artifacts/temporary-artifact.tgz, key: temporary-artifact.txt"
	
2022-12-02 11:32:23	
time="2022-12-02T09:32:23.061Z" level=info msg="/var/run/argo/outputs/artifacts/tmp/temporary-artifact.txt.tgz -> /tmp/argo/outputs/artifacts/temporary-artifact.tgz"
	
2022-12-02 11:32:23	
time="2022-12-02T09:32:23.061Z" level=info msg="Copying /tmp/temporary-artifact.txt from container base image layer to /tmp/argo/outputs/artifacts/temporary-artifact.tgz"
	
2022-12-02 11:32:23	
time="2022-12-02T09:32:23.061Z" level=info msg="Staging artifact: temporary-artifact"
	
2022-12-02 11:32:23	
time="2022-12-02T09:32:23.061Z" level=info msg="Saving output artifacts"
	
2022-12-02 11:32:23	
time="2022-12-02T09:32:23.061Z" level=info msg="No output parameters"
	
2022-12-02 11:32:23	
time="2022-12-02T09:32:23.061Z" level=info msg="No Script output reference in workflow. Capturing script output ignored"
	
2022-12-02 11:32:23	
time="2022-12-02T09:32:23.061Z" level=info msg="Main container completed" error="<nil>"
	
2022-12-02 11:32:15	
time="2022-12-02T09:32:15.058Z" level=info msg="Starting deadline monitor"
	
2022-12-02 11:32:15	
time="2022-12-02T09:32:15.056Z" level=info msg="Executor initialized" deadline="2022-12-02 13:31:18 +0000 UTC" includeScriptOutput=false namespace=argo podName=test-wft2-cddwq-main-2556220558 template="{\"name\":\"main\",\"inputs\":{},\"outputs\":{\"artifacts\":[{\"name\":\"temporary-artifact\",\"path\":\"/tmp/temporary-artifact.txt\",\"s3\":{\"key\":\"temporary-artifact.txt\"}}]},\"metadata\":{},\"container\":{\"name\":\"\",\"image\":\"argoproj/argosay:v2\",\"command\":[\"sh\",\"-c\"],\"args\":[\"echo \\\"can throw this away\\\" \\u003e /tmp/temporary-artifact.txt\\necho \\\"keep this\\\" \\u003e /tmp/keep-this.txt\\n\"],\"resources\":{}},\"archiveLocation\":{\"archiveLogs\":false,\"s3\":{\"endpoint\":\"s3.amazonaws.com\",\"bucket\":\"<REDACTED>\",\"region\":\"<REDACTED>\",\"insecure\":false,\"accessKeySecret\":{\"key\":\"accesskey\"},\"secretKeySecret\":{\"key\":\"secretkey\"},\"useSDKCreds\":true,\"key\":\"test-wft2-cddwq/test-wft2-cddwq-main-2556220558\"}},\"retryStrategy\":{\"limit\":\"1\",\"retryPolicy\":\"OnTransientError\"}}" version="&Version{Version:v3.4.4,BuildDate:2022-11-29T16:49:53Z,GitCommit:3b2626ff900aff2424c086a51af5929fb0b2d7e5,GitTag:v3.4.4,GitTreeState:clean,GoVersion:go1.18.8,Compiler:gc,Platform:linux/amd64,}"
	
2022-12-02 11:32:15	
time="2022-12-02T09:32:15.056Z" level=info msg="Using executor retry strategy" Duration=1s Factor=1.6 Jitter=0.5 Steps=5
	
2022-12-02 11:32:15	
time="2022-12-02T09:32:15.052Z" level=info msg="Starting Workflow Executor" version=v3.4.4
```

**Comment by appellod** [comment]
I don't want to resurrect a closed issue, but is there any workaround for this in the meantime while we wait for the fix in 3.4.5?

**Comment by juliev0** [comment]
> I don't want to resurrect a closed issue, but is there any workaround for this in the meantime while we wait for the fix in 3.4.5?

@sarabala1979 Do you anticipate we can release 3.4.5 maybe sometime this week?

**Comment by juliev0** [comment]
> I don't want to resurrect a closed issue, but is there any workaround for this in the meantime while we wait for the fix in 3.4.5?

I spoke to @sarabala1979 and he is going to try to put out 3.4.5 today - if it doesn't happen in the next few days feel free to remind here!

**Comment by heybronson** [comment]
Bump for release 3.4.5 :)

**Comment by juliev0** [comment]
> Bump for release 3.4.5 :)

Sorry for the delay. He’s saying he plans to this coming week
